### PR TITLE
Add uc 08 implementing part1

### PR DESCRIPTION
**Pull request by tobiaskleiner**

### CODE REVIEW

File: simtools/applications/production_generate_simulation_config.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+Example
+-------
+To run the simulation configuration, execute the script as follows:
+
+.. code-block:: console
+
+    simtools-production-generate-simulation-config --azimuth 60.0 --elevation 45.0 \
+      --nsb 0.3 --ctao_data_level "A" --science_case "high_precision" \
+      --file_path "path/to/fits_file.fits" --file_type "On-source" \
```
Comment (GernotMaier) [inline simtools/applications/production_generate_simulation_config.py:39]: `On-source` as file type, while the docstring above lists other options (point-like/cone)

**Comment by GernotMaier** [comment]
Running the files in the test resources directory gives me:

```
simtools-production-generate-simulation-config --azimuth 60.0 --elevation 45.0 --nsb 0.3 --ctao_data_level "A" --science_case "high_precision" --file_path tests/resources/production_dl2_fits/prod6_LaPalma-20deg_gamma_cone.N.Am-4LSTs09MSTs_ID0_reduced.fits --file_type cone --metrics_file tests/resources/production_simulation_config_metrics.yaml --site North
INFO::calculate_statistical_errors_grid_point(l132)::load_data_from_file::Unique azimuths: [0.] deg
INFO::calculate_statistical_errors_grid_point(l133)::load_data_from_file::Unique zeniths: [70.] deg
INFO::calculate_statistical_errors_grid_point(l136)::load_data_from_file::Setting initial grid point with azimuth: 0.0 deg zenith: 70.0 deg
INFO::calculate_statistical_errors_grid_point(l379)::calculate_metrics::Effective Area Error (max in validity range): 0.999999, Reference: 0.020
INFO::calculate_statistical_errors_grid_point(l324)::calculate_error_energy_estimate_bdt_reg_tree::Calculating Energy Resolution Error
INFO::calculate_statistical_errors_grid_point(l392)::calculate_metrics::Energy Estimate Error: 0.184, Reference: 0.050
INFO::production_generate_simulation_config(l149)::main::Simulation configuration: {'core_scatter_area': np.float64(200070.0), 'viewcone': np.float64(1053.0), 'number_of_events': '27499935690566.387 ct'}
INFO::production_generate_simulation_config(l152)::main::Simulation configuration saved to: /Users/maierg/DESYCloud/CTA/Pipelines/MC-Pipeline/gammasim-tools/simtools/simtools-output/production_generate_simulation_config/configured_simulation_params.json
```

What is a viewcone of 1053? I assume that is not in degrees.

Fur a future implementation, we need to discuss how we want to do the rounding for the number of events to be simulated. We of course can simulate ''27499935690566" events, but my preference would be e.g., round this to 275..

The output says `Unique zeniths: [70.] deg`. Are we sure that this is not a mixup of zenith and elevation?

File: simtools/applications/production_generate_simulation_config.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        "--science_case", type=str, required=True, help="Science case for the simulation."
+    )
+    config.parser.add_argument(
+        "--file_path", type=str, required=True, help="Path to MC event file in DL2 format."
+    )
+    config.parser.add_argument(
+        "--file_type",
+        type=str,
+        required=True,
+        help="Type of the dl2_mc_events_file file ('point-like' or 'cone').",
```
Comment (GernotMaier) [inline simtools/applications/production_generate_simulation_config.py:87]: Same as above - write DL2 MC event file.

File: simtools/model/camera.py
Comment (GernotMaier) [inline simtools/model/camera.py:1]: Not sure I understand the changes. Are they related to the configuration? A bit confused.

File: simtools/production_configuration/calculate_statistical_errors_grid_point.py [hunk 3]
```diff
@@ -0,0 +1,458 @@
+"""
+Provides functionality to evaluate statistical errors from dl2_mc_events_file files.
```
Comment (GernotMaier) [inline simtools/production_configuration/calculate_statistical_errors_grid_point.py:2]: DL2 MC Event list files. Please change this throughout this file in docstrings.

Better would also to talk about uncertainties, not errors. But let's leave it.

File: simtools/production_configuration/calculate_statistical_errors_grid_point.py [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+        """
+        self.file_path = file_path
+        self.file_type = file_type
+        self.metrics = metrics
+        self.grid_point = grid_point
+
+        self.data = self.load_data_from_file()
+
+        self.uncertainty_effective_area = None
+        self.error_energy_estimate_bdt_reg_tree = None
```
Comment (GernotMaier) [inline simtools/production_configuration/calculate_statistical_errors_grid_point.py:66]: We said we rename this so that the energy reconstruction method is not part of the name. That energies are reconstructed using BDT trees is not relevant in this context.

Name it `energy_estimate` (here and throughout the code)

File: simtools/production_configuration/calculate_statistical_errors_grid_point.py [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+                        )
+                        self.grid_point = (
+                            1 * u.TeV,
+                            unique_azimuths[0],
+                            unique_zeniths[0],
+                            0,
+                            0 * u.deg,
+                        )  # Initialize grid point with azimuth and zenith
+                    else:
+                        _logger.warning(
```
Comment (GernotMaier) [inline simtools/production_configuration/calculate_statistical_errors_grid_point.py:148]: I thought we agreed that this should lead to an error?

File: simtools/production_configuration/calculate_statistical_errors_grid_point.py
Comment (GernotMaier) [inline simtools/production_configuration/calculate_statistical_errors_grid_point.py:1]: There are a couple of lines without unit tests. Would of course be best if tests would be added.

File: simtools/production_configuration/generate_simulation_config.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
+"""
+Module defines the `SimulationConfig` class.
```
Comment (GernotMaier) [inline simtools/production_configuration/generate_simulation_config.py:2]: The short docstring should describe the functionality of this calls in one line. e.g., `Configures and generates simulation parameters for a specific grid point`

File: simtools/production_configuration/event_scaler.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
+"""
+Module defines the EventScaler class.
```
Comment (GernotMaier) [inline simtools/production_configuration/event_scaler.py:2]: Short description should describe the functionality in one line.

File: simtools/production_configuration/generate_simulation_config.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+        Returns
+        -------
+        float
+            The area factor.
+        """
+        azimuth = self.grid_point.get("azimuth", 0)
+        elevation = self.grid_point.get("elevation", 0)
+        night_sky_background = self.grid_point.get("night_sky_background", 0)
+
+        # Implement reading of factor from LUT
```
Comment (GernotMaier) [inline simtools/production_configuration/generate_simulation_config.py:151]: This should be a TODO

File: simtools/production_configuration/production_configuration_helper_functions.py
Comment (GernotMaier) [inline simtools/production_configuration/production_configuration_helper_functions.py:1]: I think this module is not needed. It is a two line expression and the functionality is given by the collect_data_from_file function. I suggest to remove it.

File: simtools/visualization/plot_camera.py
Comment (GernotMaier) [inline simtools/visualization/plot_camera.py:1]: This again is unrelated?

File: simtools/production_configuration/interpolation_handler.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
+"""Interpolates between instances of StatisticalErrorEvaluator using EventScaler."""
+
+import astropy.units as u
+import matplotlib.pyplot as plt
```
Comment (GernotMaier) [inline simtools/production_configuration/interpolation_handler.py:4]: I do suggest to separate plotting functions into their own module.

importing matplotlib is slow and it is a large package. Not ideal to import it each time we want to run interpolation.

File: simtools/production_configuration/interpolation_handler.py [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+        query_points : np.ndarray
+            Array of query points with shape (n, 5), where n is the number of points,
+            and 5 represents (energy, azimuth, zenith, nsb, offset).
+
+        Returns
+        -------
+        np.ndarray
+            Interpolated values at the query points.
+        """
+        # Remove flat dimensions from the normalized grid and query points
```
Comment (GernotMaier) [inline simtools/production_configuration/interpolation_handler.py:104]: No need for this comment. `remove_flat_dimensions` is a good function name and explains its functionality.

**Review by GernotMaier — CHANGES_REQUESTED** [review]
Thanks for this Tobias!

I think we are almost there, I have a couple of minor suggestions (mostly naming), which should not take a long time.

As discussed in person:

- we accept the low coverage for now as we expect notable code changes
- we are aware of a couple of placeholder functions, which require implementation as one of the next stages.

File: docs/changes/1098.feature.md
[the remark this answers is not in the document]
Comment (tobiaskleiner) [reply docs/changes/1098.feature.md:1]: This is from the initial PR number. Changing to 1244.

File: simtools/applications/production_generate_simulation_config.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+nsb (float, required)
+    Night sky background value.
+ctao_data_level (str, required)
+    The data level for the simulation (e.g., 'A', 'B', 'C').
+science_case (str, required)
+    The science case for the simulation.
+file_path (str, required)
+    Path to file with MC events at CTAO DL2 data level. Used for statistical uncertainty evaluation.
+file_type (str, required)
+    Type of the dl2_mc_events_file file ('point-like' or 'cone').
```
[the remark this answers is not in the document]
Comment (tobiaskleiner) [reply simtools/applications/production_generate_simulation_config.py:25]: This is what you suggested to call it in one of your last comments.

File: simtools/applications/production_generate_simulation_config.py [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+site (str, required)
+    The observatory site (North or South).
+
+Example
+-------
+To run the simulation configuration, execute the script as follows:
+
+.. code-block:: console
+
+    simtools-production-generate-simulation-config --azimuth 60.0 --elevation 45.0 \
```
[the remark this answers is not in the document]
Comment (tobiaskleiner) [reply simtools/applications/production_generate_simulation_config.py:37]: have added the path now.

File: simtools/model/camera.py
Comment (tobiaskleiner) [reply simtools/model/camera.py:1]: This is just a simple (unrelated) renaming to be consistent in the code.

File: simtools/production_configuration/interpolation_handler.py (hunk 9 above)
Comment (tobiaskleiner) [reply simtools/production_configuration/interpolation_handler.py]: Moved the import to the plotting function. Let me know if this is also fine.

File: simtools/production_configuration/interpolation_handler.py (hunk 10 above)
Comment (tobiaskleiner) [reply simtools/production_configuration/interpolation_handler.py]: fixed

**Comment by tobiaskleiner** [comment]
> Running the files in the test resources directory gives me:
> 
> ```
> simtools-production-generate-simulation-config --azimuth 60.0 --elevation 45.0 --nsb 0.3 --ctao_data_level "A" --science_case "high_precision" --file_path tests/resources/production_dl2_fits/prod6_LaPalma-20deg_gamma_cone.N.Am-4LSTs09MSTs_ID0_reduced.fits --file_type cone --metrics_file tests/resources/production_simulation_config_metrics.yaml --site North
> INFO::calculate_statistical_errors_grid_point(l132)::load_data_from_file::Unique azimuths: [0.] deg
> INFO::calculate_statistical_errors_grid_point(l133)::load_data_from_file::Unique zeniths: [70.] deg
> INFO::calculate_statistical_errors_grid_point(l136)::load_data_from_file::Setting initial grid point with azimuth: 0.0 deg zenith: 70.0 deg
> INFO::calculate_statistical_errors_grid_point(l379)::calculate_metrics::Effective Area Error (max in validity range): 0.999999, Reference: 0.020
> INFO::calculate_statistical_errors_grid_point(l324)::calculate_error_energy_estimate_bdt_reg_tree::Calculating Energy Resolution Error
> INFO::calculate_statistical_errors_grid_point(l392)::calculate_metrics::Energy Estimate Error: 0.184, Reference: 0.050
> INFO::production_generate_simulation_config(l149)::main::Simulation configuration: {'core_scatter_area': np.float64(200070.0), 'viewcone': np.float64(1053.0), 'number_of_events': '27499935690566.387 ct'}
> INFO::production_generate_simulation_config(l152)::main::Simulation configuration saved to: /Users/maierg/DESYCloud/CTA/Pipelines/MC-Pipeline/gammasim-tools/simtools/simtools-output/production_generate_simulation_config/configured_simulation_params.json
> ```
> 
> What is a viewcone of 1053? I assume that is not in degrees.
> 
> Fur a future implementation, we need to discuss how we want to do the rounding for the number of events to be simulated. We of course can simulate ''27499935690566" events, but my preference would be e.g., round this to 275..
> 
> The output says `Unique zeniths: [70.] deg`. Are we sure that this is not a mixup of zenith and elevation?

Thanks for the review @GernotMaier. Yes rounding would not be an issue. Should also think about if we should do a refit with a power-law function to the scaled events, as there is probably no possibility to do a simulation without a power-law.

File: simtools/applications/production_generate_simulation_config.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/applications/production_generate_simulation_config.py]: Yes, the name is good but not the way it is written. Docstrings are for humans, and it should be written without "_"s. Not `dl2_mc_event_file` but `DL2 MC event file`.

File: simtools/production_configuration/interpolation_handler.py (hunk 9 above)
Comment (GernotMaier) [reply simtools/production_configuration/interpolation_handler.py]: That is only the 2nd best solution, as it requires to disable a pylint rule (and we should minimize disabling them). Fine for now, but should fix this in future (throughout the code).

**Comment by tobiaskleiner** [comment]
@GernotMaier thanks again for the review. Fixed the remaining points and it should be ready to merge now. Let me know if there is anything missing.

**Comment by GernotMaier** [comment]
All good, thanks! Please merge.

**Review by GernotMaier — APPROVED** [review]
Forgot that I need to approve!
